optimum/neuron/generation/utils.py

```diff
@@ -14,8 +14,1262 @@
+          `constraints!=None` or `force_words_ids!=None`
+
+    You do not need to call any of the above methods directly. Pass custom parameter values to 'generate' instead. To
+    learn more about decoding strategies refer to the [text generation strategies guide](../generation_strategies).
+    """
+
+    def _update_model_kwargs_for_xla_generation(
+            self,
+            outputs: ModelOutput,
+            model_kwargs: Dict[str, Any],
+            batch_size: int,
+            is_encoder_decoder: bool = False,
+            standardize_cache_format: bool = False,
+            max_length: int = None,
+            seq_length: int = None,
+            use_cache=True,
+    ) -> Dict[str, Any]:
+        def _initialize_attention(model_kwargs, num_padding_values, is_encoder_decoder):
+            """initializes the appropriate attention mask -- encoder-decoder models use `decoder_attention_mask`"""
+            if is_encoder_decoder:
+                # One 1 for decoder_start_token_id, 0s for the currently-unfilled locations in the past_key_values tensor,
+                # 1s for the actual input_ids
+                decoder_attention_mask = torch.cat(
+                    [
+                        torch.zeros((batch_size, num_padding_values), dtype=torch.int32),
+                        torch.ones((batch_size, 2), dtype=torch.int32),
+                    ],
+                    axis=1,
+                ).to(outputs.logits.device)
+                mask = {"decoder_attention_mask": decoder_attention_mask}
+            else:
+                attention_mask = model_kwargs.pop("attention_mask")
+                # 0s for the currently-unfilled locations in the past_key_values tensor, 1s for the actual input_ids
+                attention_mask = torch.cat(
+                    [
+                        torch.zeros(
+                            (batch_size, num_padding_values), dtype=attention_mask.dtype, device=attention_mask.device
+                        ),
+                        attention_mask,
+                        torch.ones((batch_size, 1), dtype=attention_mask.dtype, device=attention_mask.device),
+                    ],
+                    axis=1,
+                )
+                mask = {"attention_mask": attention_mask}
+
+            return mask
+
+        def _update_attention(model_kwargs, is_encoder_decoder):
+            """updates the appropriate attention mask -- encoder-decoder models use `decoder_attention_mask`"""
+            if is_encoder_decoder:
+                decoder_attention_mask = model_kwargs.pop("decoder_attention_mask")
+                decoder_attention_mask_update_slice = torch.ones(
+                    (batch_size, 1), dtype=decoder_attention_mask.dtype, device=decoder_attention_mask.device
+                )
+                decoder_attention_mask = torch.cat(
+                    [decoder_attention_mask[:, 1:], decoder_attention_mask_update_slice], dim=-1
+                )
```

Comment:
So we remove the first token, and append a one in the end?
Just making sure I understand.